### PR TITLE
Inertial sensor driver 

### PR DESCRIPTION
**Pull request by Mirish**
The following is the driver for LSM6DS3 Inertial sensor to support the 96boards STM32 Sensor mezzanine board with Qualcomm DragonBoard 410c Board running AP. 
This is the second driver we submit as part of the porting of the AP code to DB410c. 
It was written and designed with the sensors data sheet and application note. 

[LSM6DS3.pdf](https://github.com/ArduPilot/ardupilot/files/1325201/LSM6DS3
[LSM6DS3_Application_note.pdf] 

(https://github.com/ArduPilot/ardupilot/files/1325213/LSM6DS3_Application_note.pdf)
.pdf) - datasheet

### CODE REVIEW

**Comment by OXINARF** [comment]
Hi @Mirish, thank you for your first contribution! Please read our contribution rules (http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html), specially about the commit messages format. Also, your code isn't formatted according to our style guide (http://ardupilot.org/dev/docs/style-guide.html).
Since this is your first contribution, we can fix it for you, but I just want to learn for your next pull requests.

**Comment by Mirish** [comment]
Hi @OXINARF  thank you for your help and input, I will follow those guidelines the next time I will perform a pull request. Sorry for the inconvenience.

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    // start the timer process to read samples
+    _dev->register_periodic_callback(1250, FUNCTOR_BIND_MEMBER(&AP_InertialSensor_LSM6DS3::_poll_data, void));
+
+}
+
+
+uint8_t AP_InertialSensor_LSM6DS3::_register_read(uint8_t reg)
+{
+    uint8_t val = 0;
+	if (_sem->take_nonblocking())
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: why are you taking the frontend semaphore for a register read?

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+	if (_sem->take_nonblocking()) 
+	{
+    	_dev->read_registers(reg, &val, 1);
+    }
+	return val;
+}
+
+void AP_InertialSensor_LSM6DS3::_register_write(uint8_t reg, uint8_t val, bool checked)
+{
+	if (_sem->take_nonblocking())
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: again, why take this semaphore?
also, every sem take should be matches to a semaphore release

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    // convert to G/LSB to (m/s/s)/LSB */
+    _accel_scale *= GRAVITY_MSS;
+}
+
+/**
+ * Timer process to poll for new data from the LSM6DS3.
+ */
+void AP_InertialSensor_LSM6DS3::_poll_data()
+{
+    if (_gyro_data_ready())
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: gyro_data_ready and accel_data_ready both read the status register. So calling it twice just wastes time.
You should be trying to minimise the number of bus calls you are doing.

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+{
+    uint8_t status = _register_read(STATUS_REG);
+    return status & G_DATA_READY_MASK;
+}
+
+void AP_InertialSensor_LSM6DS3::_read_data_transaction_a()
+{
+    struct sensor_raw_data raw_data = {};
+    const uint8_t reg = OUT_X_L_A;
+    if (!_dev->transfer(&reg, 1, (uint8_t *) &raw_data, sizeof(raw_data)))
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: can you use the sensors FIFO instead? That is generally a lot more efficient and avoids some nasty timing issues

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	if (_sem->take_nonblocking()) 
+	{
+    	_dev->write_register(reg, val);
+	}
+}
+
+void AP_InertialSensor_LSM6DS3::_gyro_init()
+{
+    _register_write(CTRL10_C,CTRL10_C_AXIS_VALUES);
+    _register_write(CTRL2_G,0x8C);
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: what is 0x8C ?

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.h [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+AP_HAL::OwnPtr<AP_HAL::I2CDevice> _dev;
+
+    /*
+     * If data-ready GPIO pins numbers are not defined (i.e. any negative
+     * value), the fallback approach used is to check if there's new data ready
+     * by reading the status register. It is *strongly* recommended to use
+     * data-ready GPIO pins for performance reasons.
+     */
+    AP_HAL::DigitalSource * _drdy_pin_a;
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.h]: what are these drdy pin definitions for? They don't seem to be used

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp (hunk 1 above)
Comment (Mirish) [reply libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: I'm very sorry, this driver was tested, and it's values were verified to be correct , but in the hit of the moment those semaphores were added without much thought, and without testing, and make no sense at all. This kind of mistake will not be repeated.

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp (hunk 5 above)
Comment (Mirish) [reply libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: Sets to high performance (1.66 kHz) and  2000 dps. 
Define in the code was added.

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.h (hunk 6 above)
Comment (Mirish) [reply libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.h]: At first they were meant to be used and then forgot to remove them. They are removed now.

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp (hunk 3 above)
Comment (Mirish) [reply libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: Fixed by reading status register outside the function

File: libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp (hunk 4 above)
Comment (Mirish) [reply libraries/AP_InertialSensor/AP_InertialSensor_LSM6DS3.cpp]: To use the sensors FIFO we should enable it in the FIFO_CTRL5 register, I don't completely follow what should be changed in those lines.

**Comment by Mirish** [comment]
Hi @tridge, have you had a chance to look at the fixes? 
Thanks ,
Miri
:-)

**Comment by tridge** [comment]
> Hi @tridge, have you had a chance to look at the fixes?

I've had a look, and I still think it needs quite a lot of work

first off, have you actually tried it? It isn't clear to me from your PR if you have test-flown with this driver.
Next, why have separate functions for reading the gyro and accel? You should be able to do a single transfer for both. Bus transactions are expensive, so why do two? You can do a single transaction that fetches status, temperature, gyro and accel. That is why the sensor has all these registers with neighboring addresses, so you can do a single transaction.
Next, I really do think you should study the code for how to use the FIFO. You need to do the following:
enable the FIFO

- setup the FIFO control registers to control what goes into the FIFO
- on each timer callback, read the fifo status registers so you know how much data is available
- then read the right number of bytes from the FIFO
- process all those bytes

See the invensense driver for an example of how this is done.
If the FIFO doesn't work for some reason then we can do it without the FIFO (eg. the LSM303D has hardware bugs that prevent us from using the FIFO), but if it does work it will produce much better results, as it means you can't miss a sample.

**Comment by khancyr** [comment]
I will close this PR as it is stalling from long time. So changes were asked but not done, and the API heavily change since them. I would be easier to start from scratch.
Feel free to open another PR to support this sensor. Thanks

**Pull request closed without merging**